pkg/controller/gcppubsub/reconcile.go

```diff
@@ -0,0 +1,326 @@
+
+package gcppubsub
+
+import (
+	"context"
+	"fmt"
+
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+
+	"github.com/knative/eventing-sources/pkg/controller/sinks"
+
+	"cloud.google.com/go/pubsub"
+	"github.com/knative/eventing-sources/pkg/apis/sources/v1alpha1"
+	"github.com/knative/pkg/logging"
+	"go.uber.org/zap"
+	"k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/client-go/dynamic"
+	"k8s.io/client-go/rest"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const (
+	finalizerName = controllerAgentName
+)
+
+// gcpPubSubClientCreator creates a real GCP PubSub client. It should always be used, except during
+// unit tests.
+func gcpPubSubClientCreator(ctx context.Context, googleCloudProject string) (pubSubClient, error) {
+	// Auth to GCP is handled by having the GOOGLE_APPLICATION_CREDENTIALS environment variable
+	// pointing at a credential file.
+	psc, err := pubsub.NewClient(ctx, googleCloudProject)
+	if err != nil {
+		return nil, err
+	}
+	return &realGcpPubSubClient{
+		client: psc,
+	}, nil
+}
+
+type reconciler struct {
+	client        client.Client
+	dynamicClient dynamic.Interface
+	scheme        *runtime.Scheme
+
+	pubSubClientCreator pubSubClientCreator
+
+	receiveAdapterImage              string
+	receiveAdapterServiceAccountName string
+}
+
+func (r *reconciler) InjectClient(c client.Client) error {
+	r.client = c
+	return nil
+}
+
+func (r *reconciler) InjectConfig(c *rest.Config) error {
+	var err error
+	r.dynamicClient, err = dynamic.NewForConfig(c)
+	return err
+}
+
+func (r *reconciler) Reconcile(ctx context.Context, object runtime.Object) (runtime.Object, error) {
+	logger := logging.FromContext(ctx).Desugar()
+
+	src, ok := object.(*v1alpha1.GcpPubSubSource)
+	if !ok {
+		logger.Error("could not find GcpPubSub source", zap.Any("object", object))
+		return object, nil
+	}
+
+	// This Source attempts to reconcile three things.
+	// 1. Determine the sink's URI.
+	//     - Nothing to delete.
+	// 2. Create a receive adapter in the form of a Deployment.
+	//     - Will be garbage collected by K8s when this GcpPubSubSource is deleted.
+	// 3. Register that receive adapter as a Pull endpoint for the specified GCP PubSub Topic.
+	//     - This needs to deregister during deletion.
+	// Because there is something that must happen during deletion, we add this controller as a
+	// finalizer to every GcpPubSubSource.
+
+	// See if the source has been deleted.
+	deletionTimestamp := src.DeletionTimestamp
+	if deletionTimestamp != nil {
+		err := r.deleteSubscription(ctx, src)
+		if err != nil {
+			logger.Error("Unable to delete the Subscription", zap.Error(err))
+			return src, err
+		}
+		r.removeFinalizer(src)
+		return src, nil
+	}
+
+	r.addFinalizer(src)
```

Comment:
only add the finalizer after the subscription is created.

Reply:
I don't have a strong feeling on this, but it seems safer to me to add the finalizer right before the `r.createSubscription` call. That way the controller always gets a chance to undo the creation of external resources, regardless of what happens afterward.

Reply:
I agree with @grantr's line of thinking. In my mind our two options are:

1. Add the finalizer before creating the subscription.
    - Pro: Never leaks PubSub Subscriptions
1. Add the finalizer after creating the subscription.
    - Pro: GcpPubSubSources are deletable if the GCP creds don't work (as the subscription was never successfully created, so the finalizer wasn't added).

My preference is for 1, but it does come at a cost.

Reply:
I tend to agree with grantr on this. but I don't feel super stongly either, just seems cleaner.

Reply:
My other argument in favor of adding the finalizer first is that it makes it less likely we accidentally do something bad in the future. E.g. in the future we create a PubSub `Topic` before creating the `Subscription`. We would need to move adding the finalizer to before creating the `Subscription`. If we forgot, then it would be very easy to leak `Topic`s.